### PR TITLE
[search] Text index refactoring.

### PR DESCRIPTION
**Pull request by mpimenov**

### CODE REVIEW

File: search/base/text_index/reader.hpp [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+      last += ReadVarUint<uint32_t>(source);
+      fn(last);
+    }
+  }
+
+  template <typename Fn>
+  void ForEachPosting(strings::UniString const & token, Fn && fn) const
+  {
+    auto const utf8s = strings::ToUtf8(token);
+    ForEachPosting(std::move(utf8s), std::forward<Fn>(fn));
```
Comment (tatiana-yan) [inline search/base/text_index/reader.hpp]: include utility

File: search/base/text_index/dictionary.hpp [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    if (it == m_tokens.cend() || *it != token)
+      return false;
+    id = ::base::checked_cast<uint32_t>(std::distance(m_tokens.cbegin(), it));
+    return true;
+  }
+
+  void SetTokens(std::vector<Token> && tokens)
+  {
+    ASSERT(std::is_sorted(tokens.begin(), tokens.end()), ());
+    m_tokens = std::move(tokens);
```
Comment (tatiana-yan) [inline search/base/text_index/dictionary.hpp]: include utility

File: search/base/text_index/mem.hpp [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+      return;
+    for (auto const p : it->second)
+      fn(p);
+  }
+
+  template <typename Fn>
+  void ForEachPosting(strings::UniString const & token, Fn && fn) const
+  {
+    auto const utf8s = strings::ToUtf8(token);
+    ForEachPosting(std::move(utf8s), std::forward<Fn>(fn));
```
Comment (tatiana-yan) [inline search/base/text_index/mem.hpp]: include utility

File: search/base/text_index/reader.hpp (hunk 1 above)
Comment (mpimenov) [reply search/base/text_index/reader.hpp]: done

File: search/base/text_index/dictionary.hpp (hunk 2 above)
Comment (mpimenov) [reply search/base/text_index/dictionary.hpp]: done

File: search/base/text_index/mem.hpp [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+        postings.emplace_back(last);
+      }
+      CHECK_EQUAL(source.Pos(), postingsStarts[i + 1], ());
+
+      m_postingsByToken.emplace(tokens[i], postings);
+    }
+  }
+
+  template <typename Sink>
+  static uint32_t RelativePos(Sink & sink, uint64_t startPos)
```
[the remark this answers is not in the document]
Comment (mpimenov) [reply search/base/text_index/mem.hpp]: done

File: search/base/text_index/mem.hpp (hunk 3 above)
Comment (mpimenov) [reply search/base/text_index/mem.hpp]: done

**Comment by mpimenov** [comment]
@tatiana-yan PTAL

**Comment by mpimenov** [comment]
JTTP

**Pull request merged**